### ISSUE TITLE
Cannot respond to plugging or unplugging of touchscreen devices during NVDA operation.

### ISSUE THREAD

**Issue by hwf1324** [nvaccess/nvda#15197]
### Steps to reproduce:

1. Run NVDA.
2. When plugging in a touchscreen device, disconnect the touchscreen and vice versa.
3. Press NVDA+1 to enter the help mode to view touch interaction gestures, and to see if the Touch Interaction category in the settings dialog exists.

### Actual behavior:

1. When a touch screen device is present, unplug this device after starting NVDA. The Touch Interaction panel still exists in the setup dialogue box. After modifying the configuration items in this panel, you cannot exit the setup dialogue box with the OK button.
2. When no touch screen device exists, plug in the touch screen device after starting NVDA. The NVDA must be restarted for the NVDA to load the touch screen function.

### Expected behavior:

1. the Setup dialogue box removes the Touch Interaction panel after unplugging all touch screen devices while NVDA is running.
2. When there is no touch screen device, during NVDA operation, plug in the touch screen device and NVDA loads the Touch Interaction function in time.

#### NVDA installed/portable/running from source:

installed

#### NVDA version:

2023.1

#### Windows version:

Windows 11 Version 22H2 (OS Build 22621.1992)

#### Does the issue still occur after restarting your computer?

Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

No

**Comment by hwf1324** [comment]
I have a few initial thoughts on this issue.

1. Detect the USB device plug event and run the touch screen detection function when triggered.
2. detect the presence of a touchscreen device each time a setup dialogue is created.

However, this issue has little impact, consider whether to add runtime overhead code for this?

**Comment by hwf1324** [comment]
I have a few initial thoughts on this issue.

1. Detect the USB device plug event and run the touch screen detection function when triggered.
2. detect the presence of a touchscreen device each time a setup dialogue is created.

However, this issue has little impact, consider whether to add runtime overhead code for this?

**Comment by josephsl** [comment]
Hi, what happens when you try all this with Narrator? Thanks.

**Comment by hwf1324** [comment]
Hi @josephsl , thanks for your reply.

> Hi, what happens when you try all this with Narrator? Thanks.

After my testing, the narrator responds correctly to the plugging and unplugging of the touch screen device.

1. restart the computer when there is no touch screen device.
2. Start the narrator.
3. Plug in the touchscreen device.

The narrator responds to my touch gestures.

**Comment by josephsl** [comment]
Hi, I see. Is this a USB powered touchscreen monitor? If yes, then I think it would be better to listen to a hardware event that informs Windows and applications that displays have changed. That way we can work with various ports such as USB, HDMI, and others. Thanks.

**Comment by josephsl** [comment]
Hi, I see. @MichaelDCurran, do you know if touch handler threads can respond to hardware changes? I know we can start and stop touch handler itself, but not sure if it can handle hardware changes. If yes, then I think we could come up with something unless it can’t be done. Thanks.

**Comment by josephsl** [comment]
Hi, regarding settings panel: I don’t think it can be done easily as touch interaction settings panel is located between document formatting and Windows OCR panels. This is because touchscreen support is available in Windows 8.x and later whereas Windows OCR panel is available on Windows 10 and later. When NVDA supports Windows 10 and later only, that should make it easier to insert and remove touch interaction panel. Thanks.

**Comment by hwf1324** [comment]
> Hi, regarding settings panel: I don’t think it can be done easily as touch interaction settings panel is located between document formatting and Windows OCR panels. This is because touchscreen support is available in Windows 8.x and later whereas Windows OCR panel is available on Windows 10 and later. When NVDA supports Windows 10 and later only, that should make it easier to insert and remove touch interaction panel. Thanks.

I wonder if it's feasible to hide the touch interaction category when judging `windll.user32.GetSystemMetrics(SM_MAXIMUMTOUCHES)` to be 0 every time I show the settings dialog?

**Comment by josephsl** [comment]
Hi, not really as the list of settings panels (including touch interaction) is initialized at NVDA startup. Thanks.

**Comment by CyrilleB79** [comment]
> Hi, not really as the list of settings panels (including touch interaction) is initialized at NVDA startup. Thanks.

@josephsl, inserting a panel in the middle of the list while NVDA is running does not seem a problem to me. Panels are removed and added again (at the end of the list) when add-ons are reloaded with NVDA+control+F3. I have not checked it but it should be possible to look for the index of "Document navigation" panel and insert a panel just after it.

**Comment by hwf1324** [comment]
> @josephsl, inserting a panel in the middle of the list while NVDA is running does not seem a problem to me. Panels are removed and added again (at the end of the list) when add-ons are reloaded with NVDA+control+F3. I have not checked it but it should be possible to look for the index of "Document navigation" panel and insert a panel just after it.

@CyrilleB79, The issue does not seem to be inserting panels at runtime, but rather dynamically removing panels at runtime. This is because changing the settings in the Touch Interaction panel after unplugging the touchscreen device results in not being able to exit the settings dialogue box with the OK button.

The best way to do this is to be able to hide the items in catListCtrl. This would restore access to the touch interaction panel when this device is re-inserted.

But this treats the symptoms rather than the root cause.

**Comment by Brian1Gaff** [comment]
I never noticed as I just disabled it in the settings of windows as it was giving me a headache. 
 I guess you only encounter this if you visit the  dialogue with a device that was unplugged, rather than disabled. Why would you go there if you had removed it?
 Brian

-- 
***@***.*** 
Sent via blueyonder.(Virgin media) 
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.
